### ISSUE TITLE
Do not convert 'undefined' and 'null' values in HttpParams contructor

### ISSUE THREAD

**Issue by xripcsu** [angular/angular#46272]
### Which @angular/* package(s) are relevant/related to the feature request?

common

### Description

In previous versions of angular I was able to remove undefined and null values from HttpParams Object in interceptor.
After upgrading to version 14 I am no longer able to remove them because all parameters are now converted to string.

### Proposed solution

Add exception for null and undefined values so it is possible to handle them in interceptor.

### Alternatives considered

-

**Comment by AndrewKushnir** [comment]
Related to #42643.

**Comment by AndrewKushnir** [comment]
@xripcsu could you please share a bit more information about your use-case and create a [Stackblitz](https://stackblitz.com/) demo, so that we can look at it?

**Comment by xripcsu** [comment]
Here is version 13 https://stackblitz.com/edit/angular-ivy-iisqap?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp.module.ts 

And here is version 14 https://stackblitz.com/edit/angular-ivy-fovqek?file=src%2Fapp%2Fapp.component.ts,package.json,src%2Fapp%2Fapp.module.ts

So now because all values are converted to strings I am not able to tell which parameters were originally 'null' or 'undefined'. Because of that I can not use interceptor to remove those parameters from requests.

**Comment by Triskae** [comment]
Hello,
Exact same issue here, workaround is to test 'undefined' or 'null' instead of undefined or null, but not really clean ...

```
        params.get(key) === null ||
        params.get(key) === undefined ||
        params.get(key) === 'undefined' ||
        params.get(key) === 'null' 
 ```

**Comment by xripcsu** [comment]
yes I was thinking about that but this would make it impossible to use 'undefined' or 'null' string values

**Comment by Triskae** [comment]
Aaaah yes, right ! 
In our case we don't need undefined nor null, but yeah, if you need those values, this workaround is not the right one.

**Comment by xripcsu** [comment]
Well most of my parameters are coming from form groups which were not strongly typed before, and for inputs like dropdown it is common to have undefined as value (at least when material dropdown is used).

**Comment by terencehonles** [comment]
I had previously used an `HttpUrlEncodingCodec` in order to handle this issue for our application, but that's applied too late now as they are all converted to strings and I had to work around this change in another way that would intercept the `HttpParams` before it is called.

We are gradually increasing the type safety of our code, but we've got plenty of room to grow.

Might it make sense to not convert the values to string is if it is null or undefined and change the typing on the params and encoding functions to accept those types. The default implementation could do what it is doing right now, but this would allow an `HttpUrlEncodingCodec` override to handle this case specifically how an application needs to?
